frontend/src/fileReceiver.js

```diff
@@ -61,60 +61,51 @@ class FileReceiver extends EventEmitter {
         true,
         ['encrypt', 'decrypt']
       )
-    ])
-      .then(([fdata, key]) => {
-        this.emit('decrypting', true);
-        return Promise.all([
-          window.crypto.subtle
-            .decrypt(
-              {
-                name: 'AES-GCM',
-                iv: hexToArray(fdata.iv),
-                additionalData: hexToArray(fdata.aad)
-              },
-              key,
-              fdata.data
-            )
-            .then(decrypted => {
-              this.emit('decrypting', false);
-              return new Promise((resolve, reject) => {
-                resolve(decrypted);
-              });
-            }),
-          new Promise((resolve, reject) => {
-            resolve(fdata.filename);
-          }),
-          new Promise((resolve, reject) => {
-            resolve(hexToArray(fdata.aad));
+    ]).then(([fdata, key]) => {
+      this.emit('decrypting', true);
+      return Promise.all([
+        window.crypto.subtle.decrypt(
+          {
+            name: 'AES-GCM',
+            iv: hexToArray(fdata.iv),
+            additionalData: hexToArray(fdata.aad)
+          },
+          key,
+          fdata.data
+        ).then(decrypted => {
+          this.emit('decrypting', false);
+          return new Promise((resolve, reject) => {
+            resolve(decrypted);
           })
-        ]);
+        }),
+        new Promise((resolve, reject) => {
+          resolve(fdata.filename);
+        }),
+        new Promise((resolve, reject) => {
+          resolve(hexToArray(fdata.aad));
+        })
+      ]);
+    }).then(([decrypted, fname, proposedHash]) => {
+      this.emit('hashing', true);
+      return window.crypto.subtle.digest('SHA-256', decrypted).then(calculatedHash => {
+        this.emit('hashing', false);
+        const integrity = new Uint8Array(calculatedHash).toString() === proposedHash.toString();
+        if (!integrity) {
+          this.emit('unsafe', true)
+          return Promise.reject();
+        } else {
+          this.emit('safe', true);
+          return Promise.all([
+            new Promise((resolve, reject) => {
+              resolve(decrypted);
+            }),
+            new Promise((resolve, reject) => {
+              resolve(fname);
+            })
```

Comment:
Same here. I think you can just do `return Promise.all([decrypted, fname]);` or possibly even `Promise.resolve([decrypted, fname]);`.